design_documents/test_design.md

```diff
@@ -0,0 +1,30 @@
+# Test design
+
+## Context
+
+Current test design in Starknet Forge has some issues:
+- gas estimation is problematic since we have to distinguish between pure Cairo gas
+(e.g., creating an array in test) and Starknet gas
+- starknet functions like `get_caller_address` are unavailable in tests directly
+- starknet system calls like `library_call` are unavailable in tests directly
+- testing internal contract functions requires exposing them as `external`
+- operating on the contract execution layer while tests being pure Cairo code can create confusion
+
+## Goal
+
+Propose a solution of test design that solves most of the aforementioned issues.
+
+## Proposed Solution
+
+Tests should be contracts (just like in the early days of Protostar).
+Basically, the code in test should behave as if it was a function called in an invoke transaction.
+This clears up the confusion concerning gas estimation — we just calculate the cost of the whole test 
+as it is all executed in the Starknet context. Moreover, starknet functions and system calls now can be made available
+in tests directly. Last but not least, the confusion concerning what actually happens when using dispatchers will be cleared:
```

Comment:
I think what is still unclear is that what state is used for the system calls made directly in tests. And how would one influence it.

E.g. we can allow using `get_caller_address`, but what it would return? A default value? How is it useful at all? Should be able to apply cheatcodes to the test state and how would that work?

Reply:
State of the test (as each test is a contract) will be used - basically we will have storage implementation for the test. When it comes to functions like `get_caller_address` that depend on the "who called the test" we will prob just use default values. Not sure what to do with cheatcodes - I guess it depends on the implementation that@MaksymilianDemitraszek is working on. From what I understood from our conversation offline some cheatcodes won't make any sense (like prank, roll etc.) since we just want to give a test a state (storage access) but don't want to include it in blockifier's address map